src/Analyzer/QueryAnalysisPass.cpp

```diff
@@ -0,0 +1,5340 @@
+{
+    size_t nested_path_parts_size = nested_path.getPartsSize();
+    for (size_t i = 0; i < nested_path_parts_size; ++i)
+    {
+        const auto & nested_path_part = nested_path[i];
+        auto tuple_element_function = std::make_shared<FunctionNode>("tupleElement");
+        tuple_element_function->getArguments().getNodes().push_back(expression_node);
+        tuple_element_function->getArguments().getNodes().push_back(std::make_shared<ConstantNode>(nested_path_part));
+        expression_node = tuple_element_function;
+    }
+
+    return expression_node;
+}
+
+/** Try to get lambda node from sql user defined functions if sql user defined function with function name exists.
+  * Returns lambda node if function exists, nullptr otherwise.
+  */
+QueryTreeNodePtr QueryAnalyzer::tryGetLambdaFromSQLUserDefinedFunction(const std::string & function_name)
+{
+    auto user_defined_function = UserDefinedSQLFunctionFactory::instance().tryGet(function_name);
+    if (!user_defined_function)
+        return {};
+
+    const auto & create_function_query = user_defined_function->as<ASTCreateFunctionQuery>();
+    auto result_node = buildQueryTree(create_function_query->function_core, context);
+    if (result_node->getNodeType() != QueryTreeNodeType::LAMBDA)
+        throw Exception(ErrorCodes::LOGICAL_ERROR,
+            "SQL user defined function {} must represent lambda expression. Actual {}",
+            function_name,
+            create_function_query->function_core->formatForErrorMessage());
+
+    return result_node;
+}
+
+/// Evaluate scalar subquery and perform constant folding.
+void QueryAnalyzer::evaluateScalarSubquery(QueryTreeNodePtr & node, size_t subquery_depth)
+{
+    auto * query_node = node->as<QueryNode>();
+    auto * union_node = node->as<UnionNode>();
+    if (!query_node && !union_node)
+        throw Exception(ErrorCodes::LOGICAL_ERROR,
+            "Node must have query or union type. Actual {} {}",
+            node->getNodeTypeName(),
+            node->formatASTForErrorMessage());
+
+    if ((query_node && query_node->hasConstantValue()) ||
+        (union_node && union_node->hasConstantValue()))
+        return;
+
+    auto subquery_context = Context::createCopy(context);
+    Settings subquery_settings = context->getSettings();
+    subquery_settings.max_result_rows = 1;
+    subquery_settings.extremes = false;
+    subquery_context->setSettings(subquery_settings);
+
+    if (!subquery_context->hasQueryContext())
+    {
+        // auto subquery_query_context = subquery_context->getQueryContext();
+        // for (const auto & it : data.scalars)
+        //     context->addScalar(it.first, it.second);
+    }
+
+    auto options = SelectQueryOptions(QueryProcessingStage::Complete, subquery_depth, true /*is_subquery*/);
+    auto interpreter = std::make_unique<InterpreterSelectQueryAnalyzer>(node, options, subquery_context);
+
+    auto io = interpreter->execute();
```

Comment:
It will work as before in `ExecuteScalarSubqueriesVisitor`. Cancel does not work for scalar subqueries.

Reply:
Cancel (or KILL) does work on scalar subqueries.

```
Select (Select sum(number) from system.numbers);
```

Can be killed. It does not work when using `clickhouse client` + Ctrl+C because the network isn't processed, but a second connection will do the trick.

It works fine with the old interpreter but the analyzer breaks this.

Reply:
The fix is fairly simple:

```diff
diff --git a/src/Analyzer/QueryAnalysisPass.cpp b/src/Analyzer/QueryAnalysisPass.cpp
index 04d636c5c5..2668114d47 100644
--- a/src/Analyzer/QueryAnalysisPass.cpp
+++ b/src/Analyzer/QueryAnalysisPass.cpp
@@ -1193,6 +1193,7 @@ void QueryAnalyzer::evaluateScalarSubqueryIfNeeded(QueryTreeNodePtr & node, size
     Block block;
     PullingAsyncPipelineExecutor executor(io.pipeline);
     io.pipeline.setProgressCallback(context->getProgressCallback());
+    io.pipeline.setProcessListElement(context->getProcessListElement());
     while (block.rows() == 0 && executor.pull(block))
     {
     }
```

This is covered by the `02136_kill_scalar_queries.sh` test. It doesn't pass with the analyzer as is, it passes with the change. Feel free to add it, or we can revisit and send PRs once this is merged